### PR TITLE
Overhaul code towards a more ROSic system

### LINKED ISSUE #36: rosparam move to config file
Rosparameters should live in a config file instead of in the launch file

### LINKED ISSUE #41: Undefined variable 'rmax'
This line tries to change the sign of rmax variable which is undefined.

https://github.com/nasa-jpl/osr-rover-code/blob/e26ef24b28ad5f313e4af9e9ec70c7c9cdceac79/ROS/osr/scripts/robot.py#L88

### LINKED ISSUE #53: Switch to ROS logger
ROS has its own logging system with varying log levels (debug, info, warn, error, fatal). Instead of using python prints, which cannot be filtered or display different severity levels, all print statements used in ROS nodes should be converted to rospy logs. This is done in some places already but not in all.

Instead of commenting out some debugging print statements, you can also decide to log them at the debug level, which can be conveniently turned off or filtered (e.g. rosconsole)

### LINKED ISSUE #55: osr_msgs: list --> fields
Currently the messages contain an ordered list for each motor which is hard to interpret and error prone. Either the messages should be converted to contain specific fields, e.g.:

```
drive:
  int32 left_front
  int32 left_middle
  int32 left_back
  int32 right_front
  int32 right_middle
  int32 right_back
steer:
  int32 left_front
  int32 left_back
  int32 right_front
  int32 right_back
```

OR the order should be better documented. I believe currently the order is 131M1, 131M2, 132M1, 132M2. Perhaps we want to change this to be clockwise starting front right or similar.

### LINKED ISSUE #59: Consider switching to SI units
To be consistent with the robotics and ROS community, it would make sense to switch to SI units for all calculations. See ROS' [REP103](https://www.ros.org/reps/rep-0103.html). Instead of degrees and degrees per second, we would use radian and rad/s as well as meter and m/s iso inches.

Happy to do this if there is consensus on this.

### LINKED ISSUE #60: calculateTargetDeg method clipping code runs but not used
The code for `calculateTargetDeg` contains something like:

```
ang1 = int(math.degrees(math.atan(self.d1 / (abs(r) + self.d3))))
ang2 = int(math.degrees(math.atan(self.d2 / (abs(r) + self.d3))))
ang3 = int(math.degrees(math.atan(self.d2 / (abs(r) - self.d3))))
ang4 = int(math.degrees(math.atan(self.d1 / (abs(r) - self.d3))))

angles = [ang1, ang2, ang3, ang4]
for i in range(4):
	if angles[i] < -45:
		angles[i] = -43
	elif angles[i] > 45:
		angles[i] = 43

if radius > 0:
	return [ang2, -ang1, -ang4, ang3]
else:
	return [-ang4, ang3, ang2, -ang1]
```

The middle section seems to be clipping values outside the range abs(45) to abs(43) but those values are assigned to the variables `angles`, which isn't used in the following return value calculation.

Options:

1. remove clipping, it's not functional code
2. fix the clipping if it needs to stay

### LINKED ISSUE #64: tick2deg assumes 90 deg range for corner motors
Although easy to calculate and independent of encoder model,

* this method can only be used for corner encoders as it assumes a range of 90 degrees
* the range isn't exactly 90 degrees, due to mechanical inaccuracies and soft limits.

My proposal is to convert ticks based on the constant ration tick/radian or tick/degree (related: #59) and have a setting for this ratio for those who have different encoders/motors.

### LINKED ISSUE #65: Potential divide by zero  should be wrapped in a Try Except block
Depending on the setting's wheel distance you can cause a divide by zero error when updating position, breaking the normal joystick behavior. This can be easily remedied with a Try-Catch block or a check for equality
https://github.com/nasa-jpl/osr-rover-code/blob/e26ef24b28ad5f313e4af9e9ec70c7c9cdceac79/ROS/osr/scripts/robot.py#L128-L129

### PR DESCRIPTION
**Pull request by Achllle**
## Description

Quite large overhaul of the code. Many improvements, tracked through the following issues:

* close #65
* close #64 
* close #60
* close #59 
* close #55 
* close #53 
* close #41 
* close #36 
* greatly simplify math on corner and drive motor velocity/position calculation
* improve code quality

## Dependencies

- [x] #52 
- [x] #57

### CODE REVIEW

File: ROS/osr_bringup/config/physical_properties.yaml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+rover_dimensions:
+  d1: 0.184  # [m, ~7.254in]
+  d2: 0.267  # [m, ~10.5in]
+  d3: 0.267  # [m, ~10.5in]
+  d4: 0.2556  # [m, ~10.073in]
+  wheel_radius: 0.075  # [m]
+drive_no_load_rpm: 130  # no load speed for the drive motors
```
Comment (Achllle) [inline ROS/osr_bringup/config/physical_properties.yaml]: Needs to be adjusted

**Comment by ericjunkins** [comment]
@Achllle , got test testing and verification of this today. Ran into one issue with the calculation of the encoder turns. I think what is best is to merge this in, and I'll open a new issue and PR to deal with that so we can get all the code here in, and get things tracked with a bit better granularity.

In general this all is very well done and super clean, really good job on all of it! 👍

**Comment by Achllle** [comment]
Normally what should happen is that we fix it here - you always want master in working condition. I guess we can make an exception since master wasn't really working properly. The comment I made above needs to be adjusted since that is a value that only is true for the drive motors I use. I'll make a new issue for that.

**Comment by mikcox** [comment]
I wanted to second @ericjunkins comment. You've been doing excellent work and thank you so much for your contributions to this project! You're making it better for everyone down the line who builds it.

**Comment by leonardosent** [comment]
Thanks @Achllle for letting me know about the corner steering issue with my rover. I want to update my ROS code since it is so out of date. Should I grab what’s in the main branch or wait a bit longer until the remaining bugs are fixed?

**Comment by Achllle** [comment]
Happy to help and continue working on this. I appreciate the recognition!

@leonardosent You can grab what's on master, simply `git checkout master` and `git pull origin master`. If it doesn't work for some reason, you can always go back to where you were before with `git checkout branch-or-commit-hash-where-you-want-to-be`. Let us know if something isn't working for you (by opening an issue) and we'll fix it ASAP. It would be good to have more folks test out the new code.

**Pull request merged**